### PR TITLE
Add bindings to CUDA tensors and storages in THPP

### PR DESCRIPTION
**Pull request by colesbury**

### CODE REVIEW

File: torch/lib/THPP/Tensor.hpp [hunk 1]
```diff
@@ -36,6 +36,8 @@ struct Tensor {
   virtual bool isContiguous() const = 0;
   virtual void* data() = 0;
   virtual const void* data() const = 0;
+  virtual void* cdata() = 0;
+  virtual const void* cdata() const = 0;
```
Comment (apaszke) [inline torch/lib/THPP/Tensor.hpp]: Is it really needed? The whole point of this class is to abstract the tensor type, and you'll have to cast this pointer to a concrete class afterwards. Why not `dynamic_cast<THTensor*>(t)->cdata()` (i.e. implement `cdata` only on concrete classes)?

File: torch/lib/THPP/Traits.hpp [hunk 2]
```diff
@@ -12,77 +12,89 @@
 
 namespace thpp {
 
+template<typename T>
+struct type_traits {};
+
 template<typename real>
 struct interface_traits {
   using scalar_type = typename std::conditional<
-    std::is_floating_point<real>::value,
```
Comment (apaszke) [inline torch/lib/THPP/Traits.hpp]: What's wrong with that? half?

File: torch/lib/THPP/storages/THCStorage.hpp [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  virtual THCStorage& fast_set(std::size_t ind, scalar_type value) override;
+  virtual scalar_type get(std::size_t ind) override;
+  virtual scalar_type fast_get(std::size_t ind) override;
+
+  virtual thpp::Type type() const override;
+  virtual bool isCuda() const override;
+  virtual int getDevice() const override;
+
+  virtual std::unique_ptr<Tensor> newTensor() const override;
+  virtual storage_type *getRaw() const;
```
Comment (apaszke) [inline torch/lib/THPP/storages/THCStorage.hpp]: This is equivalent to `cdata`. Leave only one of them.

**Review by apaszke — CHANGES_REQUESTED** [review]

File: torch/lib/THPP/Tensor.hpp (hunk 1 above)
Comment (colesbury) [inline torch/lib/THPP/Tensor.hpp]: Not always. For example, if you interface with the cuDNN bindings, you need a THVoidTensor*

File: torch/lib/THPP/Traits.hpp (hunk 2 above)
Comment (colesbury) [inline torch/lib/THPP/Traits.hpp]: It's not a floating point type. It's a struct containing a short.

File: torch/lib/THPP/storages/THCStorage.hpp (hunk 3 above)
Comment (colesbury) [inline torch/lib/THPP/storages/THCStorage.hpp]: They're different in their return types. `getRaw()` has the behavior you described above, while `cdata()` always returns `void*`. You can't override a function that returns `void*` with one that returns a pointer to a class.

File: torch/lib/THPP/Tensor.hpp (hunk 1 above)
Comment (apaszke) [inline torch/lib/THPP/Tensor.hpp]: Yeah but that's such an ugly hack 😕 We could rewrite them to use THPP.

Comment (colesbury) [inline torch/lib/THPP/Tensor.hpp]: Yeah I was hoping to do that eventually, but someone didn't want to review a 3,000 line diff :P 

I'd like to get things Conv and BatchNorm working with the autograd refactor before changing the cuDNN bindings.

Comment (apaszke) [inline torch/lib/THPP/Tensor.hpp]: Hehe. Ok we can leave it 😊

**Review by apaszke — APPROVED** [review]

**Pull request merged**